allennlp/models/language_model.py

```diff
@@ -0,0 +1,206 @@
+
+from allennlp.training.metrics.perplexity import Perplexity
+
+@Model.register("word-lm")
+class WordLM(Model):
+    """
+    This ``WordLM`` is a word level language model.
+
+    Parameters
+    ----------
+    vocab : ``Vocabulary``, required
+        A Vocabulary, required in order to compute sizes for input/output projections.
+    text_field_embedder : ``TextFieldEmbedder``, required
+        Used to embed the ``tokens`` ``TextField`` we get as input to the model.
+    encoder : ``LMRNN``, required
+        The Recurrent Neural Networks for language modeling.
+    softmax : ``AdaptiveSoftmax``
+        The adaptive softmax for predicting next words
+    proj: bool, optional (default=``True``)
+        whether add a linear projection before the softmax layer
+    relu: bool, optional (default=``True``)
+        whether add a ReLU non-linear transformation before the softmax layer
+    initializer : ``InitializerApplicator``, optional (default=``InitializerApplicator()``)
+        Used to initialize the model parameters.
+    regularizer : ``RegularizerApplicator``, optional (default=``None``)
+        If provided, will be used to calculate the regularization penalty during training.
+    """
+    def __init__(self, vocab: Vocabulary,
+                 text_field_embedder: TextFieldEmbedder,
+                 encoder: Seq2SeqEncoder,
+                 softmax: SoftmaxWithNLL,
+                 proj: bool = False,
+                 relu: bool = False,
+                 dropout: float = 0.0,
+                 batch_first: bool=True,
+                 initializer: InitializerApplicator = InitializerApplicator(),
+                 regularizer: Optional[RegularizerApplicator] = None) -> None:
+
+        super().__init__(vocab, regularizer)
+
+        self.encoder = encoder
+        self.batch_first = batch_first
+        self.text_field_embedder = text_field_embedder
+
+        if dropout:
+            self.dropout = torch.nn.Dropout(dropout)
+            dropout_list = [self.dropout]
+        else:
+            self.dropout = None
+            dropout_list = []
+
+        if proj:
+            self.proj = nn.Linear(self.encoder.get_output_dim(), softmax.get_input_dim())
+            proj_list = [self.proj] 
+            if relu:
+                proj_list += [nn.ReLU()]
+            proj_list = proj_list + dropout_list
+        else:
+            self.proj = None
+            proj_list = []
+
+        after_encoder_list = dropout_list + proj_list
+        if len(after_encoder_list) > 0:
+            self.after_encoder = nn.Sequential(*after_encoder_list)
+        else:
+            self.after_encoder = None
+
+        self.softmax = softmax
+        if not self.softmax.adaptive and self.softmax.head.weight.size() == self.text_field_embedder.token_embedder_tokens.weight.size():
+            self.softmax.head.weight = self.text_field_embedder.token_embedder_tokens.weight
+
+        self.softmax_in = softmax.get_input_dim()
```

Comment:
rename self.softmax_in to something with `size` or `dim` in the name to help disambiguate from other softmax input related objects (tensors, etc)